src/mrdsrg-helper/dsrg_mem.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Forte: an open-source plugin to Psi4 (https://github.com/psi4/psi4)
+ * that implements a variety of quantum chemistry methods for strongly
+ * correlated electrons.
+ *
+ * Copyright (c) 2012-2020 by its authors (see COPYING, COPYING.LESSER, AUTHORS).
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ *
+ * @END LICENSE
+ */
+
+#ifndef _dsrg_mem_h_
+#define _dsrg_mem_h_
+
+#include <vector>
+#include <string>
+#include <map>
+
+#include "base_classes/mo_space_info.h"
+
+namespace forte {
+
```

Comment:
What does this class do?

Reply:
This class is created to check/print the memory usage before any real DSRG computation. Different DSRG levels have different memory requirement and it is useful to terminate the job if it lacks of memory.

Basically the user sets the max memory available and a map of the space character to its size (e.g., 'c': 20, 'a': 4, 'v': 80, 'h': 24, 'p': 84, 'g': 104).

Take LDSRG(2) as an example. It stores F, V, T1, T2, Hbar1, Hbar2, and other intermediates. To compute the memory usage of T1 and T2, the user can `add_entry` and pass the name and the labels: `add_entry("T1 and T2 amplitudes", {"hp", "hhpp"});`.

There are two sets of global intermediates used in LDSRG(2) both of sizes g^2 + g^4 and this can be added as `add_entry("Global intermediates", {"gg", "gggg"}, 2);`.

Besides global memory usage, there are also local usage. In this case, the max local memory usage will be the bottleneck. To add a local memory usage, use `add_entry("Local storage for S1 and S2 amplitudes", {"hp", "hhpp"}, 1, false);`

The `print` function will print all passed in entries and terminate the job if the global memory plus the max local memory is greater than the max available set by the user.